chrome/content/zotero/xpcom/itemPaneManager.js

```diff
@@ -0,0 +1,358 @@
+ * @property {(options: SectionInitHookArgs) => void} [onInit]
+ * Lifecycle hook called when section is initialized
+ *
+ * Do:
+ * 1. Initialize data if necessary
+ * 2. Setup hooks, e.g. notifier callback
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionBasicHookArgs) => void} [onDestroy]
+ * Lifecycle hook called when section is destroyed
+ *
+ * Do:
+ * 1. Remove data and release resource
+ * 2. Remove hooks, e.g. notifier callback
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionDataChangeHookArgs) => boolean} [onDataChange]
+ * Lifecycle hook called when section incoming data change received
+ *
+ * Return `false` to abort the data change, otherwise the data change will be applied
+ * (stored internally and can be accessed by `options.getData` in other hooks).
+ *
+ * Do:
+ * 1. Validate and decide whether to accept the incoming data change
+ * 2. Update the section enabled state with `options.setEnabled`. For example, if the section
+ *   is only enabled in the readers, you can use:
+ * ```js
+ * onDataChange(options) {
+ *   let { incomingData, setEnabled } = options.;
+ *   if (incomingData.type === "tabType") {
+ *     setEnabled(incomingData.value === "reader");
+ *   }
+ *   return true;
+ * }
+ * ```
+ * options.setEnabled(options.tabType === "reader");
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionHookArgs) => void | Promise<void>} onRender
+ * Lifecycle hook called when section should do primary render
+ *
+ * Create elements and append them to `options.body`.
+ * - If `onRender` is sync, it is always called when the item pane is rendered;
+ * - If `onRender` is async, it will be called when scrolling into view.
+ *
+ * If the rendering can be slow, you should either make it async, or move async part to `onSecondaryRender`.
+ *
+ * > Note that the rendering of section is fully controlled by Zotero to minimize resource usage.
+ * > Only render UI things when you are told to.
+ * @property {(options: SectionHookArgs) => void | Promise<void>} [onSecondaryRender]
+ * [Optional] Lifecycle hook called when section should do secondary render
+ *
+ * The best practice to time-consuming rendering with runtime decided section height is:
+ * 1. Compute height and create a box in sync `onRender`;
+ * 2. Render actual contents in async `onSecondaryRender`.
+ * @property {(options: SectionEventHookArgs) => void} [onToggle] - Called when section is toggled
+ * @property {SectionButton[]} [sectionButtons] - Section button options
+ */
+
+
+class ItemPaneManager {
+	_customSections = {};
+
+	_lastUpdateTime = 0;
+
+	/**
+	 * Register a custom section in item pane. All registered sections must be valid, and must have a unique paneID.
+	 * @param {ItemDetailsSectionOptions | ItemDetailsSectionOptions[]} options - section data
+	 * @returns {string | string[] | false} - The paneID or false if no section were added
+	 */
+	registerSections(options) {
```

Comment:
Why should this be `registerSections` (plural)? The vast majority of the time, won't people be creating a single section? If we really thought people register multiple would be common, we could have a `registerSections` as well, but we should definitely have a `registerSection` for the common case.

Reply:
I agree we don't need to support registering multiple sections! The behavior is inherited from item tree API, though I don't think it's necessary here. I'll update it